torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+        self.global_ref_rank = -1  # Either the destination or the src rank, if reducing or broadcasting for instance
+        self.global_rank = -1
+
+    def unroll(self) -> None:
+        """
+        Dsitribute the contents of the flat buffer back to the attached parameters
+        """
+
+        for flat in self.params:
+            if self.global_ref_rank != self.global_rank and self.gradients_based:
+                # this rank is not the owner, release the grad
+                flat.param.grad = None
+            else:
+                flat.param.data.copy_(self.buffer[flat.start : flat.stop].view_as(flat.param.data), non_blocking=True)
+
+        self.reset()
+
+    def reset(self) -> None:
+        """ empty the bucket """
+        self.current_offset = 0
+        self.params.clear()
+
+    def append(self, tensor: torch.Tensor) -> bool:
+        """ add a tensor to the bucket """
+
+        end = self.current_offset + tensor.numel()
+
+        if end > self.max_size:
+            return False
+
+        data_source = tensor.grad.data if use_gradient else tensor.data  # type: ignore    # mypy is drunk
+        self.buffer[self.current_offset : end].copy_(data_source.view(-1))
+        self.params.append(FlatParam(tensor=tensor, start=self.current_offset, stop=end))
+        self.current_offset = end
+        return True
+
+    def full(self) -> bool:
+        """ is the bucket full ? """
+        return self.current_offset == self.max_offset
+
+
+class ZeroRedundancyOptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ZeroRedundancyOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters at each rank.
+    Each parameter belongs to a single rank and is not divided among ranks.
+    The partition is arbitrary and does not correspond to the information flow for instance.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
+    Arguments:
+        params (list of tensors):
+            parameters to be optimized
+    Keyword Args:
+        optim (torch.nn.Optimizer):
+            optimizer to shard
+        group (group):
+            torch.distributed group (default: group.WORLD)
+        bucket_cap_kb (int):
+            the size of the buffer used to batch the small parameter tensors, in kilobytes (default 128k).
+    """
+
+    def __init__(
+        self,
+        params: _params_t,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        bucket_cap_kb: int = 2 ** 17,
+        **default: Any,
+    ):
+        # Hold all the model params in the root .param_groups
+        self.in_super_constructor = True
```

Comment:
the flag would still need to be defined twice here, because the super constructor calls add_param_group, so apart from a false/true flip it does not look like a big change, unless I'm missing something ?